source/scripts/Background/controllers/DAppController.ts

```diff
@@ -0,0 +1,190 @@
+}
+interface ISigRequest {
+  address: string;
+  message: string;
+  origin: string;
+}
+
+const DAppController = (): IDAppController => {
+  let current: IDAppInfo = { origin: '', logo: '', title: '' };
+  let request: ISigRequest;
+
+  const isDAppConnected = (origin: string) => {
+    const { whitelist } = store.getState().dapp;
+
+    return !!whitelist[origin as keyof IDAppState];
+  };
+
+  const fromPageConnectDApp = (origin: string, title: string) => {
+    current = {
+      origin,
+      logo: `chrome://favicon/size/64@1x/${origin}`,
+      title,
+    };
+
+    return isDAppConnected(origin);
+  };
+
+  const fromUserConnectDApp = (
+    origin: string,
+    dapp: IDAppInfo,
+    network: string,
+    accounts: string[]
+  ) => {
+    store.dispatch(listNewDapp({ id: origin, dapp, network, accounts }));
+  };
+
+  const _dispatchEvents = async (events: any[]) => {
+    const background = await browser.runtime.getBackgroundPage();
+
+    events.forEach((event) => background.dispatchEvent(event));
+  };
+
+  const notifyAccountsChanged = async (accounts: string[]): Promise<void> => {
+    const state = store.getState();
+    const { whitelist, listening } = state.dapp;
+
+    const events: any[] = [];
+
+    // Will only notify whitelisted dapps that are listening for a wallet change.
+    for (const origin of Object.keys(listening)) {
+      const site = whitelist[origin as any];
+      const listeningEvents = listening[origin];
+
+      if (!listeningEvents.includes('accountsChanged')) {
+        return;
+      }
+
+      if (site) {
+        console.log('accounts', accounts);
+        // const siteAccounts = site.accounts as IDappAccounts;
+        // const allAccountsWithDuplicates = accounts.concat(
+        //   siteAccounts.Syscoin,
+        //   siteAccounts.Ethereum
+        // );
+        // const matchingAccounts = filter(
+        //   allAccountsWithDuplicates,
+        //   (value, index, iteratee) =>
+        //     includes(iteratee, value as string, index + 1)
+        // );
+        // if (matchingAccounts.length) {
+        //   const ethAccounts = matchingAccounts.filter((account) =>
+        //     account.toLowerCase().startsWith('0x')
+        //   );
+        //   const sysAccounts = matchingAccounts.filter((account) =>
+        //     account.toLowerCase().startsWith('sys')
+        //   );
+        //   // Dispatch a separate event for each chain
+        //   const _events = [
+        //     new CustomEvent('accountsChanged', {
+        //       detail: { data: ethAccounts, origin, chain: 'ethereum' },
+        //     }),
+        //     new CustomEvent('accountsChanged', {
+        //       detail: { data: sysAccounts, origin, chain: 'syscoin' },
+        //     }),
+        //   ];
+        //   events = [...events, ..._events];
+        // }
+      }
+    }
+
+    return _dispatchEvents(events);
+  };
+
+  const notifySiteDisconnected = async (origin: string): Promise<void> => {
```

Comment:
Mark as private.
Maybe rename to `notifySiteDisconnection` or `notifyDisconnectedSite`